From a2971b023709ea7335964abe5ddab93b6e70b74f Mon Sep 17 00:00:00 2001
From: Cedric Roijakkers <croijakkers@ilionx.com>
Date: Tue, 26 Nov 2024 09:06:22 +0100
Subject: [PATCH] Patch for kernel 6.12

---
 module/evdi_drm_drv.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/module/evdi_drm_drv.c b/module/evdi_drm_drv.c
index fb6b6a8..13e674b 100644
--- a/module/evdi_drm_drv.c
+++ b/module/evdi_drm_drv.c
@@ -73,6 +73,10 @@ static const struct file_operations evdi_driver_fops = {
 #endif
 
 	.llseek = noop_llseek,
+
+#if defined(FOP_UNSIGNED_OFFSET)
+	.fop_flags = FOP_UNSIGNED_OFFSET,
+#endif
 };
 
 #if KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)
-- 
2.43.5

