From 934e31ec29ca7abcc8fcc1d6104214885c56acf0 Mon Sep 17 00:00:00 2001
From: Jakub Prussak <jakub.prussak@synaptics.com>
Date: Fri, 20 Sep 2024 13:17:17 +0200
Subject: [PATCH] Align with Linux v6.11+

---
 module/evdi_modeset.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/module/evdi_modeset.c b/module/evdi_modeset.c
index dfa9d10..7da0532 100644
--- a/module/evdi_modeset.c
+++ b/module/evdi_modeset.c
@@ -504,7 +504,10 @@ static int evdi_crtc_init(struct drm_device *dev)
 
 static const struct drm_mode_config_funcs evdi_mode_funcs = {
 	.fb_create = evdi_fb_user_fb_create,
+#if KERNEL_VERSION(6, 11, 0) < LINUX_VERSION_CODE || defined(EL8)
+#else
 	.output_poll_changed = NULL,
+#endif
 	.atomic_commit = drm_atomic_helper_commit,
 	.atomic_check = drm_atomic_helper_check
 };
