From a2971b023709ea7335964abe5ddab93b6e70b74f Mon Sep 17 00:00:00 2001
From: Cedric Roijakkers <croijakkers@ilionx.com>
Date: Tue, 26 Nov 2024 09:06:22 +0100
Subject: [PATCH 04/31] Patch for kernel 6.12

---
 module/evdi_drm_drv.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/module/evdi_drm_drv.c b/module/evdi_drm_drv.c
index fb6b6a8..13e674b 100644
--- a/module/evdi_drm_drv.c
+++ b/module/evdi_drm_drv.c
@@ -73,6 +73,10 @@ static const struct file_operations evdi_driver_fops = {
 #endif
 
 	.llseek = noop_llseek,
+
+#if defined(FOP_UNSIGNED_OFFSET)
+	.fop_flags = FOP_UNSIGNED_OFFSET,
+#endif
 };
 
 #if KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)

From 3651b6debf631febf470106d43199d7fbd7bfd56 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Tue, 10 Dec 2024 13:52:58 +0100
Subject: [PATCH 05/31] Fix build for 6.14-rc3

---
 module/evdi_gem.c     | 2 +-
 module/evdi_painter.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/module/evdi_gem.c b/module/evdi_gem.c
index 06a032c..a8d1279 100644
--- a/module/evdi_gem.c
+++ b/module/evdi_gem.c
@@ -29,7 +29,7 @@
 #include <linux/vmalloc.h>
 
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS(DMA_BUF);
+MODULE_IMPORT_NS("DMA_BUF");
 #endif
 
 #if KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)
diff --git a/module/evdi_painter.c b/module/evdi_painter.c
index b748484..72fc79d 100644
--- a/module/evdi_painter.c
+++ b/module/evdi_painter.c
@@ -36,7 +36,7 @@
 #endif
 
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS(DMA_BUF);
+MODULE_IMPORT_NS("DMA_BUF");
 #endif
 
 #if KERNEL_VERSION(5, 1, 0) <= LINUX_VERSION_CODE || defined(EL8)

From e28e466b4fe7205f9478d43ad554d9a5e734c121 Mon Sep 17 00:00:00 2001
From: Tj <hacker@iam.tj>
Date: Thu, 8 Feb 2024 10:33:00 +0000
Subject: [PATCH 06/31] module: add README with instructions on usage

---
 module/README.md | 48 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)
 create mode 100644 module/README.md

diff --git a/module/README.md b/module/README.md
new file mode 100644
index 0000000..fc2578c
--- /dev/null
+++ b/module/README.md
@@ -0,0 +1,48 @@
+# evdi kernel module
+When installing DisplayLink software using the provided installer or via distro-provided scripts some required configuration is written. When working with the module manually that configuration will not be created automatically. This is likely to cause failure of evdi to discover external DisplayLink devices and lead to much confusion.
+
+See the shell script `dkms_install.sh` and its `evdi_add_mod_options()` function for the steps required.
+
+In summary:
+
+  1. Add module options (nmost importantly `initial_device_count=` )
+  2. Auto-load the module
+
+## Module options
+Typically `/etc/modprobe.d/evdi.conf` will contain two lines, the second being generated and system-specific:
+```
+options evdi initial_device_count=4
+softdep evdi pre: drm_kms_helper i915
+```
+Note: `softdep` above shows the host has an integrated Intel GPU (`i915`).
+
+## Auto-loading the module
+The  `dkms_install.sh` script assumes the host is using systemd and adds the auto-load config in `/etc/modules-load.d/` that is used by `systemd-modules-load.service`. For hosts without systemd the convention for loading modules at start-up is via an entry in `/etc/modules`. `modules-load.d` usually has a symbolic link to this file.
+
+Adding `evdi` to `/etc/modules` therefore should work for a broader range of possible init system scenarios.
+
+## Usage
+It is not obvious that `evdi` (as of v1.14.1) does not auto-detect and configure DisplayLink devices - something one might expect it to do. For that reason the module option `initial_device_count` is required for automatic discovery at load-time.
+
+Alternatively one can manually add outputs with:
+```
+# echo 1 > /sys/devices/evdi/add
+```
+and confirm with:
+```
+$ cat /sys/devivces/evdi/count
+```
+and when in an Xorg session:
+```
+$ xrandr --listproviders
+Providers: number : 2
+Provider 0: id: 0x43 cap: 0x9, Source Output, Sink Offload crtcs: 3 outputs: 2 associated providers: 5 name:modesetting
+Provider 1: id: 0x789 cap: 0x2, Sink Output crtcs: 1 outputs: 1 associated providers: 1 name:modesetting
+```
+Here `Provider 1` is the new output and if things work correctly a new output should appear:
+```
+$ xrandr --query
+...
+DVI-I-1-1 disconnected (normal left inverted right x axis y axis)
+```
+Once a monitor is detected the resolutions will be reported and GUI tools that manage displays should update to show the newly connected display.

From f89d5b3c105b551d9282d5949e75688b38b93275 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?=
 <75313478+synaptics-lspintzyk@users.noreply.github.com>
Date: Wed, 11 Dec 2024 14:28:48 +0100
Subject: [PATCH 07/31] Update README.md with explanation of the module
 options.

---
 module/README.md | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/module/README.md b/module/README.md
index fc2578c..01b9691 100644
--- a/module/README.md
+++ b/module/README.md
@@ -1,5 +1,6 @@
 # evdi kernel module
-When installing DisplayLink software using the provided installer or via distro-provided scripts some required configuration is written. When working with the module manually that configuration will not be created automatically. This is likely to cause failure of evdi to discover external DisplayLink devices and lead to much confusion.
+When installing DisplayLink software using the provided installer or via distro-provided scripts some required configuration is written. When working with the module manually that configuration will not be created automatically.
+This is are various settings fixing issues with compositors (XServer, Gnome/Wayland) likely to cause failure of evdi to discover external DisplayLink devices and lead to much confusion.
 
 See the shell script `dkms_install.sh` and its `evdi_add_mod_options()` function for the steps required.
 
@@ -14,15 +15,16 @@ Typically `/etc/modprobe.d/evdi.conf` will contain two lines, the second being g
 options evdi initial_device_count=4
 softdep evdi pre: drm_kms_helper i915
 ```
-Note: `softdep` above shows the host has an integrated Intel GPU (`i915`).
+Note: `initial_device_count` this is workaround for XServer that builds static lists af attached gpu devices on startup and might crash when gpu device is added dynamically.
+Note: `softdep` option tells linux kernel to load evdi driver after primary gpu driver is loaded. This is to ensure that compositor will not attempt to use evdi driver as primary gpu.
 
 ## Auto-loading the module
-The  `dkms_install.sh` script assumes the host is using systemd and adds the auto-load config in `/etc/modules-load.d/` that is used by `systemd-modules-load.service`. For hosts without systemd the convention for loading modules at start-up is via an entry in `/etc/modules`. `modules-load.d` usually has a symbolic link to this file.
+The  `dkms_install.sh` adds the auto-load config in `/etc/modules-load.d/` when it exists. For hosts without systemd the convention for loading modules at start-up is via an entry in `/etc/modules`. `modules-load.d` usually has a symbolic link to this file.
 
-Adding `evdi` to `/etc/modules` therefore should work for a broader range of possible init system scenarios.
+Adding `evdi` to `/etc/modules` enables system to load the driver on boot time.
 
 ## Usage
-It is not obvious that `evdi` (as of v1.14.1) does not auto-detect and configure DisplayLink devices - something one might expect it to do. For that reason the module option `initial_device_count` is required for automatic discovery at load-time.
+Evdi driver is just kernel space driver that adds virtual displays to the Linux. DisplayLink device enumeration and control is done in binary driver that can be downloaded from https://www.synaptics.com/products/displaylink-graphics/downloads/ubuntu.
 
 Alternatively one can manually add outputs with:
 ```

From 90859a66b62027a136a2841482d65a3bd79735bd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?=
 <75313478+synaptics-lspintzyk@users.noreply.github.com>
Date: Wed, 11 Dec 2024 14:29:02 +0100
Subject: [PATCH 08/31] Fix typo

Co-authored-by: Tuur Vanhoutte <4633209+zjeffer@users.noreply.github.com>
---
 module/README.md | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/module/README.md b/module/README.md
index 01b9691..56d72e0 100644
--- a/module/README.md
+++ b/module/README.md
@@ -32,7 +32,7 @@ Alternatively one can manually add outputs with:
 ```
 and confirm with:
 ```
-$ cat /sys/devivces/evdi/count
+$ cat /sys/devices/evdi/count
 ```
 and when in an Xorg session:
 ```

From 5f4aa52f769afc9ce096a14492af322ccb330975 Mon Sep 17 00:00:00 2001
From: "Michael L. Young" <elgueromexicano@gmail.com>
Date: Sun, 8 Dec 2024 11:25:21 -0500
Subject: [PATCH 09/31] Fix building on latest EL9 kernels

This commit fixes building evdi on the latest EL9 kernels which have backported
changes from the 6.11 kernel.

While going through the code, made some other adjustments for what applies to
EL8 and what doesn't.

Signed-off-by: Michael L. Young <elgueromexicano@gmail.com>
---
 module/evdi_drm_drv.c      | 4 ++--
 module/evdi_drm_drv.h      | 2 +-
 module/evdi_fb.c           | 2 +-
 module/evdi_gem.c          | 4 ++--
 module/evdi_i2c.c          | 2 +-
 module/evdi_modeset.c      | 2 +-
 module/evdi_painter.c      | 5 +++--
 module/evdi_platform_dev.c | 1 +
 module/evdi_platform_dev.h | 1 +
 9 files changed, 13 insertions(+), 10 deletions(-)

diff --git a/module/evdi_drm_drv.c b/module/evdi_drm_drv.c
index 13e674b..7959534 100644
--- a/module/evdi_drm_drv.c
+++ b/module/evdi_drm_drv.c
@@ -34,7 +34,7 @@
 #include "evdi_debug.h"
 #include "evdi_drm.h"
 
-#if KERNEL_VERSION(6, 8, 0) <= LINUX_VERSION_CODE || defined(EL8)
+#if KERNEL_VERSION(6, 8, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #define EVDI_DRM_UNLOCKED 0
 #else
 #define EVDI_DRM_UNLOCKED DRM_UNLOCKED
@@ -130,7 +130,7 @@ static struct drm_driver driver = {
 	.fops = &evdi_driver_fops,
 
 	.gem_prime_import = drm_gem_prime_import,
-#if KERNEL_VERSION(6, 6, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 6, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #else
 	.prime_fd_to_handle = drm_gem_prime_fd_to_handle,
 	.prime_handle_to_fd = drm_gem_prime_handle_to_fd,
diff --git a/module/evdi_drm_drv.h b/module/evdi_drm_drv.h
index 92cd1a9..ebadc2b 100644
--- a/module/evdi_drm_drv.h
+++ b/module/evdi_drm_drv.h
@@ -18,7 +18,7 @@
 #include <linux/mutex.h>
 #include <linux/device.h>
 #include <linux/i2c.h>
-#if KERNEL_VERSION(5, 5, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if KERNEL_VERSION(5, 5, 0) <= LINUX_VERSION_CODE || defined(EL8)
 #include <drm/drm_drv.h>
 #include <drm/drm_fourcc.h>
 #include <drm/drm_ioctl.h>
diff --git a/module/evdi_fb.c b/module/evdi_fb.c
index 93847c4..6bd364e 100644
--- a/module/evdi_fb.c
+++ b/module/evdi_fb.c
@@ -417,7 +417,7 @@ static int evdifb_create(struct drm_fb_helper *helper,
 	info->fix.smem_len = size;
 	info->fix.smem_start = (unsigned long)efbdev->efb.obj->vmapping;
 
-#if KERNEL_VERSION(6, 4, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 4, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #elif KERNEL_VERSION(4, 20, 0) <= LINUX_VERSION_CODE || defined(EL8)
 	info->flags = FBINFO_DEFAULT;
 #else
diff --git a/module/evdi_gem.c b/module/evdi_gem.c
index a8d1279..b0e5ce1 100644
--- a/module/evdi_gem.c
+++ b/module/evdi_gem.c
@@ -184,8 +184,8 @@ int evdi_drm_gem_mmap(struct file *filp, struct vm_area_struct *vma)
 	if (ret)
 		return ret;
 
-/* Some VMA modifier function patches present in 6.3 were reverted in EL kernels */
-#if KERNEL_VERSION(6, 3, 0) <= LINUX_VERSION_CODE
+/* Some VMA modifier function patches present in 6.3 were reverted in EL8 kernels */
+#if KERNEL_VERSION(6, 3, 0) <= LINUX_VERSION_CODE || defined(EL9)
 	vm_flags_mod(vma, VM_MIXEDMAP, VM_PFNMAP);
 #else
 	vma->vm_flags &= ~VM_PFNMAP;
diff --git a/module/evdi_i2c.c b/module/evdi_i2c.c
index 5e7f7e0..1616c21 100644
--- a/module/evdi_i2c.c
+++ b/module/evdi_i2c.c
@@ -40,7 +40,7 @@ int evdi_i2c_add(struct i2c_adapter *adapter, struct device *parent,
 	void *ddev)
 {
 	adapter->owner  = THIS_MODULE;
-#if KERNEL_VERSION(6, 8, 0) <= LINUX_VERSION_CODE || defined(EL8)
+#if KERNEL_VERSION(6, 8, 0) <= LINUX_VERSION_CODE || defined(EL9)
 #else
 	adapter->class  = I2C_CLASS_DDC;
 #endif
diff --git a/module/evdi_modeset.c b/module/evdi_modeset.c
index 7da0532..cadd791 100644
--- a/module/evdi_modeset.c
+++ b/module/evdi_modeset.c
@@ -504,7 +504,7 @@ static int evdi_crtc_init(struct drm_device *dev)
 
 static const struct drm_mode_config_funcs evdi_mode_funcs = {
 	.fb_create = evdi_fb_user_fb_create,
-#if KERNEL_VERSION(6, 11, 0) < LINUX_VERSION_CODE || defined(EL8)
+#if KERNEL_VERSION(6, 11, 0) < LINUX_VERSION_CODE || defined(EL9)
 #else
 	.output_poll_changed = NULL,
 #endif
diff --git a/module/evdi_painter.c b/module/evdi_painter.c
index 72fc79d..7828cfa 100644
--- a/module/evdi_painter.c
+++ b/module/evdi_painter.c
@@ -31,10 +31,11 @@
 
 #include <linux/dma-buf.h>
 #include <linux/vt_kern.h>
-#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE || defined(EL8)
 #include <linux/compiler_attributes.h>
 #endif
 
+/* Import of DMA_BUF namespace was reverted in EL8 */
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
@@ -718,7 +719,7 @@ void evdi_painter_dpms_notify(struct evdi_painter *painter, int mode)
 	switch (mode) {
 	case DRM_MODE_DPMS_ON:
 		painter->fg_console = fg_console;
-#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 4, 0) <= LINUX_VERSION_CODE || defined(EL8)
 		fallthrough;
 #else
 #endif
diff --git a/module/evdi_platform_dev.c b/module/evdi_platform_dev.c
index f601b4c..b47f2d6 100644
--- a/module/evdi_platform_dev.c
+++ b/module/evdi_platform_dev.c
@@ -85,6 +85,7 @@ int evdi_platform_device_probe(struct platform_device *pdev)
 	return PTR_ERR_OR_ZERO(dev);
 }
 
+/* EL9 kernel removed the callback that was returning void  */
 #if KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE
 void evdi_platform_device_remove(struct platform_device *pdev)
 #else
diff --git a/module/evdi_platform_dev.h b/module/evdi_platform_dev.h
index 1324090..5dddcbb 100644
--- a/module/evdi_platform_dev.h
+++ b/module/evdi_platform_dev.h
@@ -32,6 +32,7 @@ struct platform_device *evdi_platform_dev_create(struct platform_device_info *in
 void evdi_platform_dev_destroy(struct platform_device *dev);
 
 int evdi_platform_device_probe(struct platform_device *pdev);
+/* EL9 kernel removed the callback that was returning void  */
 #if KERNEL_VERSION(6, 11, 0) <= LINUX_VERSION_CODE
 void evdi_platform_device_remove(struct platform_device *pdev);
 #else

From fb79dca31768bf6c16a8a787fdd67b245e4611f4 Mon Sep 17 00:00:00 2001
From: "Michael L. Young" <elgueromexicano@gmail.com>
Date: Wed, 11 Dec 2024 22:32:13 -0500
Subject: [PATCH 10/31] Fix building on kernels prior to 6.13

The commit 3651b6debf631febf470106d43199d7fbd7bfd56 causes an issue for kernels
prior to 6.13-rc3.

Signed-off-by: Michael L. Young <elgueromexicano@gmail.com>
---
 module/evdi_gem.c     | 2 ++
 module/evdi_painter.c | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/module/evdi_gem.c b/module/evdi_gem.c
index b0e5ce1..e381cae 100644
--- a/module/evdi_gem.c
+++ b/module/evdi_gem.c
@@ -29,6 +29,8 @@
 #include <linux/vmalloc.h>
 
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
+#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
 
diff --git a/module/evdi_painter.c b/module/evdi_painter.c
index 7828cfa..bea139a 100644
--- a/module/evdi_painter.c
+++ b/module/evdi_painter.c
@@ -37,6 +37,8 @@
 
 /* Import of DMA_BUF namespace was reverted in EL8 */
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
+#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
 

From 23fb6ab3d480da90acde653231290ede6d171746 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Thu, 12 Dec 2024 15:10:22 +0100
Subject: [PATCH 11/31] Fix build for kernel v6.13

---
 module/evdi_gem.c     | 2 +-
 module/evdi_painter.c | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/module/evdi_gem.c b/module/evdi_gem.c
index e381cae..78a5ebd 100644
--- a/module/evdi_gem.c
+++ b/module/evdi_gem.c
@@ -29,7 +29,7 @@
 #include <linux/vmalloc.h>
 
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS(DMA_BUF);
+MODULE_IMPORT_NS("DMA_BUF");
 #elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
diff --git a/module/evdi_painter.c b/module/evdi_painter.c
index bea139a..42f537d 100644
--- a/module/evdi_painter.c
+++ b/module/evdi_painter.c
@@ -37,7 +37,7 @@
 
 /* Import of DMA_BUF namespace was reverted in EL8 */
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS(DMA_BUF);
+MODULE_IMPORT_NS("DMA_BUF");
 #elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif

From 398f065074448e9f6ed4e87443c1a7250ab16c36 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Fri, 13 Dec 2024 11:32:32 +0100
Subject: [PATCH 12/31] Fix DMA_BUF import NS on kenrels < 6.13

---
 module/evdi_gem.c     | 7 ++++---
 module/evdi_painter.c | 6 +++---
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/module/evdi_gem.c b/module/evdi_gem.c
index 78a5ebd..8d82206 100644
--- a/module/evdi_gem.c
+++ b/module/evdi_gem.c
@@ -28,10 +28,11 @@
 #include <drm/drm_cache.h>
 #include <linux/vmalloc.h>
 
-#if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS("DMA_BUF");
-#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
+
+#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
+#elif KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
 #endif
 
 #if KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)
diff --git a/module/evdi_painter.c b/module/evdi_painter.c
index 42f537d..6e1bb8a 100644
--- a/module/evdi_painter.c
+++ b/module/evdi_painter.c
@@ -36,10 +36,10 @@
 #endif
 
 /* Import of DMA_BUF namespace was reverted in EL8 */
-#if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
-MODULE_IMPORT_NS("DMA_BUF");
-#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
+#elif KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
 #endif
 
 #if KERNEL_VERSION(5, 1, 0) <= LINUX_VERSION_CODE || defined(EL8)

From c40f0c7ce130dd3c2fc411aeedcd11650e3088fd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Wed, 11 Dec 2024 14:22:55 +0100
Subject: [PATCH 13/31] Create evdi.conf file under /etc/modules when
 /etc/modules-load.d directory does not exists

---
 module/dkms_install.sh | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/module/dkms_install.sh b/module/dkms_install.sh
index 74fc0d1..b8d9602 100755
--- a/module/dkms_install.sh
+++ b/module/dkms_install.sh
@@ -68,7 +68,8 @@ evdi_dkms_install()
 
 evdi_add_mod_options()
 {
-  local module_file='/etc/modules-load.d/evdi.conf'
+  local module_file='/etc/modules/evdi.conf'
+  [ -d '/etc/modules-load.d' ] && module_file='/etc/modules-load.d/evdi.conf'
   echo 'evdi' > "$module_file"
 
   local conf_file='/etc/modprobe.d/evdi.conf'

From 6d961c43adf98444b99a751569ed45ec3f33c3be Mon Sep 17 00:00:00 2001
From: Milana Kananovich <Milana.Kananovich@synaptics.com>
Date: Tue, 10 Sep 2024 11:01:17 +0200
Subject: [PATCH 14/31] remove set -e from install scripts

---
 ci/prepare_deb_package | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/ci/prepare_deb_package b/ci/prepare_deb_package
index 05be525..20e50f5 100755
--- a/ci/prepare_deb_package
+++ b/ci/prepare_deb_package
@@ -132,7 +132,9 @@ EOF
   cat > "$control_dir/postrm" << 'EOF'
 #!/bin/bash
 
-set -e
+if [ ${DISPLAYLINK_DEBUG_DEV:-"no"} == "yes" ]; then
+  set -e
+fi
 
 if [[ $1 == purge ]]; then
   rm -rf /var/log/displaylink

From 00af5e020afdd8ecb10ded654d58beb08b12e4fe Mon Sep 17 00:00:00 2001
From: Milana Kananovich <Milana.Kananovich@synaptics.com>
Date: Wed, 11 Sep 2024 13:35:26 +0200
Subject: [PATCH 15/31] auto uninstall existing dlm

---
 ci/prepare_deb_package | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/ci/prepare_deb_package b/ci/prepare_deb_package
index 20e50f5..a41180f 100755
--- a/ci/prepare_deb_package
+++ b/ci/prepare_deb_package
@@ -107,10 +107,18 @@ EOF
   cat > "$control_dir/preinst" << EOF
 #!/bin/bash
 
-if [[ -e /opt/displaylink/evdi.tar.gz ]]
-then
-  echo >&2 "Error: Another EVDI distribution detected"
-  exit 1
+displaylink_installer_exists()
+{
+  [[ $(command -v displaylink-installer) ]]
+}
+if [[ -e /opt/displaylink/evdi.tar.gz ]]; then
+  if displaylink_installer_exists; then
+    echo "evdi.tar.gz detected: uninstalling..."
+    displaylink-installer uninstall
+  else
+    echo >&2 "evdi.tar.gz detected, but displaylink-installer is absent, can't uninstall"
+    exit 1
+  fi
 fi
 EOF
 

From b12d6ff48f93bb13fb9304fe8605bf69eff5d73a Mon Sep 17 00:00:00 2001
From: Milana Kananovich <Milana.Kananovich@synaptics.com>
Date: Mon, 16 Sep 2024 18:06:14 +0200
Subject: [PATCH 16/31] change preinst part to runtime and refactor preinst

---
 ci/prepare_deb_package | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/ci/prepare_deb_package b/ci/prepare_deb_package
index a41180f..2de0053 100755
--- a/ci/prepare_deb_package
+++ b/ci/prepare_deb_package
@@ -109,14 +109,20 @@ EOF
 
 displaylink_installer_exists()
 {
-  [[ $(command -v displaylink-installer) ]]
+  [[ \$(command -v displaylink-installer) ]]
 }
-if [[ -e /opt/displaylink/evdi.tar.gz ]]; then
+
+evdi_installed_via_standalone_run_installer()
+{
+  [[ -e /opt/displaylink/evdi.tar.gz ]]
+}
+
+if evdi_installed_via_standalone_run_installer; then
   if displaylink_installer_exists; then
-    echo "evdi.tar.gz detected: uninstalling..."
+    echo "evdi installed via standalone installer: uninstalling..."
     displaylink-installer uninstall
   else
-    echo >&2 "evdi.tar.gz detected, but displaylink-installer is absent, can't uninstall"
+    echo >&2 "ERROR: evdi installed via standalone installer but displaylink-installer script does not exists."
     exit 1
   fi
 fi
@@ -140,7 +146,7 @@ EOF
   cat > "$control_dir/postrm" << 'EOF'
 #!/bin/bash
 
-if [ ${DISPLAYLINK_DEBUG_DEV:-"no"} == "yes" ]; then
+if [ \${DISPLAYLINK_DEBUG_DEV:-"no"} == "yes" ]; then
   set -e
 fi
 

From c1163d3d19b633aaebfaa67d4b7a84c24c48ec59 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Tue, 1 Oct 2024 14:37:58 +0200
Subject: [PATCH 17/31] Workaround Gnome/mutter issue where it is not waiting
 for primary gpu

Gnome on start is iterating over all /dev/dri/card devices to find connected output.
However there is timing issue and evdi is loading first but there is not amdgpu yet.
When that happens then Gnome stucks on start with black screen on built-in monitor and never recovers.

Installer detects loaded primary gpu drm modules and adds them as soft pre-dependencies of evdi module.
---
 module/dkms_install.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/module/dkms_install.sh b/module/dkms_install.sh
index b8d9602..affdc3c 100755
--- a/module/dkms_install.sh
+++ b/module/dkms_install.sh
@@ -78,7 +78,7 @@ evdi_add_mod_options()
   echo "options evdi initial_device_count=4" > "$conf_file"
 
   local drm_deps
-  drm_deps=$(sed -n '/^drm_kms_helper/p' /proc/modules | awk '{print $4}' | tr ',' ' ')
+  drm_deps=$(sed -n '/^drm_[[:alpha:]]*_helper/p' /proc/modules | awk '{print $4}' | tr ',' '\n' | sort -u | tr '\n' ' ')
   drm_deps=${drm_deps/evdi/}
   [[ -z $drm_deps ]] && return 0
 

From cd9b67bfa7311758d848d850a87ccd77bf9d32ca Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 7 Oct 2024 09:24:59 +0200
Subject: [PATCH 18/31] Fix check for available git mirror with 5 sec timeout

---
 ci/build_against_kernel | 2 +-
 ci/run_kunit            | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/ci/build_against_kernel b/ci/build_against_kernel
index e836e6c..ebd0562 100755
--- a/ci/build_against_kernel
+++ b/ci/build_against_kernel
@@ -149,7 +149,7 @@ prepare_tar_gz_sources() { # PACKAGE SOURCES CC
 clone_linux()
 {
   local linux_mirror="https://oauth2:glpat-jeZZawHtkoZ3xLHWGCJQ@gitlab/ppd_posix/linux.git"
-  if ! curl -k --output /dev/null --silent --head --fail "$linux_mirror"; then
+  if curl -k --output /dev/null --silent -m 5 --head --fail "$linux_mirror"; then
     git -c http.sslVerify=false clone "$linux_mirror"
     (cd linux; git config http.sslVerify false)
   else
diff --git a/ci/run_kunit b/ci/run_kunit
index a1f32b0..54b12ba 100755
--- a/ci/run_kunit
+++ b/ci/run_kunit
@@ -8,7 +8,7 @@ clone_linux()
 {
   echo "Clone linux repository"
   local linux_mirror="https://oauth2:glpat-jeZZawHtkoZ3xLHWGCJQ@gitlab/ppd_posix/linux.git"
-  if curl -k --output /dev/null --silent --head --fail "$linux_mirror"; then
+  if curl -k --output /dev/null --silent -m 5 --head --fail "$linux_mirror"; then
     git -c http.sslVerify=false clone --depth 1 "$linux_mirror"
     (cd linux; git config http.sslVerify false)
   else

From 88d67b63487ad484dff31d9e1355a32771787979 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 7 Oct 2024 09:17:43 +0200
Subject: [PATCH 19/31] Validate build with CONFIG_FB define

---
 ci/build_against_kernel | 2 +-
 module/evdi_fb.c        | 6 +++---
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/ci/build_against_kernel b/ci/build_against_kernel
index ebd0562..25cf174 100755
--- a/ci/build_against_kernel
+++ b/ci/build_against_kernel
@@ -236,7 +236,7 @@ build_one() { # KVER
 
   local src_dir
   local KTAG=v${KVER}
-  local extra_cflags=
+  local extra_cflags="-DCONFIG_FB"
   if [ "${KVER}" == "master" ]; then
     KTAG="origin/master"
   else
diff --git a/module/evdi_fb.c b/module/evdi_fb.c
index 6bd364e..d8feca6 100644
--- a/module/evdi_fb.c
+++ b/module/evdi_fb.c
@@ -404,7 +404,7 @@ static int evdifb_create(struct drm_fb_helper *helper,
 	fb = &efbdev->efb.base;
 
 	efbdev->helper.fb = fb;
-#if KERNEL_VERSION(6, 1, 51) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if KERNEL_VERSION(6, 2, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 	efbdev->helper.info = info;
 #else
 	efbdev->helper.fbdev = info;
@@ -464,7 +464,7 @@ static void evdi_fbdev_destroy(__always_unused struct drm_device *dev,
 {
 	struct fb_info *info;
 
-#if KERNEL_VERSION(6, 1, 51) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if KERNEL_VERSION(6, 2, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 	if (efbdev->helper.info) {
 		info = efbdev->helper.info;
 #else
@@ -556,7 +556,7 @@ void evdi_fbdev_unplug(struct drm_device *dev)
 		return;
 
 	efbdev = evdi->fbdev;
-#if KERNEL_VERSION(6, 1, 51) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if KERNEL_VERSION(6, 2, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 	if (efbdev->helper.info) {
 		struct fb_info *info;
 

From 85401c71a3f17910a42e91753b5782766d6d50c4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 28 Oct 2024 10:08:28 +0100
Subject: [PATCH 20/31] Add libevdi{abi} conflicting packages for evdi deb
 package

---
 ci/prepare_deb_package | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/ci/prepare_deb_package b/ci/prepare_deb_package
index 2de0053..5d850a3 100755
--- a/ci/prepare_deb_package
+++ b/ci/prepare_deb_package
@@ -45,7 +45,9 @@ create_evdi_deb_package()
   local deb_dir="$temp_dir/EVDI"
   local control_dir="$deb_dir/DEBIAN"
   local docs_dir="$deb_dir/usr/share/doc/$evdi_package_name"
+  local libevdi_abi
 
+  libevdi_abi=$(echo "$evdi_version" | cut -d. -f1)
   mkdir -p \
       "$control_dir" \
       "$deb_dir/usr/src" \
@@ -98,7 +100,7 @@ Maintainer: $evdi_maintainer
 Source: evdi
 Section: kernel
 Depends: mokutil, dkms, libdrm-dev, libc6-dev
-Conflicts: libevdi0, libevdi0-dev, evdi-dkms
+Conflicts: libevdi${libevdi_abi}, libevdi${libevdi_abi}-dev, evdi-dkms
 Architecture: $evdi_arch
 Priority: optional
 Description: $evdi_description

From c5481875156027c3b8367b47fd608dbd95cc8ec7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 25 Nov 2024 11:51:05 +0100
Subject: [PATCH 21/31] Test only main branch against linux kernel master

---
 Jenkinsfile_linux_master | 2 --
 1 file changed, 2 deletions(-)

diff --git a/Jenkinsfile_linux_master b/Jenkinsfile_linux_master
index 049f988..73df7dc 100644
--- a/Jenkinsfile_linux_master
+++ b/Jenkinsfile_linux_master
@@ -12,8 +12,6 @@ pipeline {
                 dir('src') {
                     script {
                       def list = [ "origin/main" ]
-                      def branches_to_build = sh(script: '''git branch -r -l origin/release/* | xargs''', returnStdout: true).trim()
-                      list.addAll(branches_to_build.split())
                       for(int i=0; i < list.size(); i++) {
                         def branchName = list[i];
                         stage(branchName) {

From 05a624fc47665da99e406d25d1af063239106a9c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Tue, 10 Dec 2024 11:31:03 +0100
Subject: [PATCH 22/31] Check github/main branch against linux master branch

---
 Jenkinsfile              | 2 ++
 Jenkinsfile_linux_master | 8 +++++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/Jenkinsfile b/Jenkinsfile
index 6eae40d..b2e1c3b 100644
--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -87,12 +87,14 @@ pipeline {
         stage ('Build against released kernels') {
             steps {
                 sh '''#!/usr/bin/env bash
+                set -e
                 ./ci/build_against_kernel --repo-ci all'''
             }
         }
         stage ('Build against latest rc kernel') {
             steps {
                 sh '''#!/usr/bin/env bash
+                set -e
                 ./ci/build_against_kernel --repo-ci rc'''
             }
         }
diff --git a/Jenkinsfile_linux_master b/Jenkinsfile_linux_master
index 73df7dc..ff7311d 100644
--- a/Jenkinsfile_linux_master
+++ b/Jenkinsfile_linux_master
@@ -11,12 +11,18 @@ pipeline {
             steps {
                 dir('src') {
                     script {
-                      def list = [ "origin/main" ]
+                     sh '''#!/usr/bin/env bash
+                       git remote add github "https://github.com/DisplayLink/evdi.git"
+                       git fetch github
+                     '''
+
+                      def list = [ "github/main", "origin/main" ]
                       for(int i=0; i < list.size(); i++) {
                         def branchName = list[i];
                         stage(branchName) {
                             withEnv(["STAGE_BRANCH=$branchName"]) {
                             sh '''#!/usr/bin/env bash
+                              set -e
                               git checkout ${STAGE_BRANCH}
                               ./ci/build_against_kernel --repo-ci master
                               ./ci/build_against_kernel --repo-ci rc

From 8b61869abb5f6baf60f538e164b3445aecc1abc6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Thu, 12 Dec 2024 07:29:48 +0100
Subject: [PATCH 23/31] Fetch checkpatch script from linux kernel mirror

---
 ci/run_style_check | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ci/run_style_check b/ci/run_style_check
index 7dae9e6..fdecc18 100755
--- a/ci/run_style_check
+++ b/ci/run_style_check
@@ -11,9 +11,10 @@ files=(
 )
 
 KVER=master
+linux_mirror="https://gitlab.com/linux-kernel/linux"
 
 for file in "${files[@]}"; do
-  wget -N -q "https://raw.githubusercontent.com/torvalds/linux/${KVER}/scripts/${file}"
+  curl -k -q "${linux_mirror}/-/raw/${KVER}/scripts/${file}" --output "${file}"
 done
 
 chmod +x checkpatch.pl

From 3a82424f1a1e0154ff6dd910a3ff15fe01ea6170 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 16 Dec 2024 09:22:57 +0000
Subject: [PATCH 24/31] Set version to 1.14.8

---
 ci/deb_config              | 2 +-
 library/Makefile           | 2 +-
 library/evdi_lib.h         | 2 +-
 module/Makefile            | 2 +-
 module/dkms.conf           | 2 +-
 module/dkms_install.sh     | 2 +-
 module/evdi_platform_drv.h | 4 ++--
 7 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/ci/deb_config b/ci/deb_config
index c7762ea..86ef144 100644
--- a/ci/deb_config
+++ b/ci/deb_config
@@ -8,7 +8,7 @@ evdi_package_name='evdi'
 evdi_maintainer='Synaptics Technical Support <technical-enquiries@synaptics.com>'
 
 # shellcheck disable=SC2034
-evdi_version='1.14.7'
+evdi_version='1.14.8'
 
 # shellcheck disable=SC2034
 evdi_description="Extensible Virtual Display Interface (EVDI) is a Linux® kernel module that enables management of multiple screens.
diff --git a/library/Makefile b/library/Makefile
index dca3f0e..0401878 100644
--- a/library/Makefile
+++ b/library/Makefile
@@ -11,7 +11,7 @@ PKG_CONFIG ?= pkg-config
 DEPS = evdi_ioctl.h
 CFLAGS := -I../module -std=gnu99 -fPIC -D_FILE_OFFSET_BITS=64 $(CFLAGS) $$($(PKG_CONFIG) --cflags-only-I libdrm)
 
-LIBVER := 1.14.7
+LIBVER := 1.14.8
 LIBABI := 1
 
 PREFIX ?= /usr/local
diff --git a/library/evdi_lib.h b/library/evdi_lib.h
index f9332af..05c3a65 100644
--- a/library/evdi_lib.h
+++ b/library/evdi_lib.h
@@ -15,7 +15,7 @@ extern "C" {
 
 #define LIBEVDI_VERSION_MAJOR 1
 #define LIBEVDI_VERSION_MINOR 14
-#define LIBEVDI_VERSION_PATCH 7
+#define LIBEVDI_VERSION_PATCH 8
 
 struct evdi_lib_version {
 	int version_major;
diff --git a/module/Makefile b/module/Makefile
index 9467938..4bbe446 100644
--- a/module/Makefile
+++ b/module/Makefile
@@ -63,7 +63,7 @@ CP ?= cp
 DKMS ?= dkms
 RM ?= rm
 
-MODVER=1.14.7
+MODVER=1.14.8
 
 ifeq ($(KVER),)
 	KVER := $(shell uname -r)
diff --git a/module/dkms.conf b/module/dkms.conf
index 83f92d8..93983bc 100644
--- a/module/dkms.conf
+++ b/module/dkms.conf
@@ -7,7 +7,7 @@
 #
 
 PACKAGE_NAME="evdi"
-PACKAGE_VERSION=1.14.7
+PACKAGE_VERSION=1.14.8
 AUTOINSTALL=yes
 
 MAKE[0]="make all INCLUDEDIR=/lib/modules/$kernelver/build/include KVERSION=$kernelver DKMS_BUILD=1"
diff --git a/module/dkms_install.sh b/module/dkms_install.sh
index affdc3c..9af412c 100755
--- a/module/dkms_install.sh
+++ b/module/dkms_install.sh
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: GPL-2.0-only
 # Copyright (c) 2023 DisplayLink (UK) Ltd.
 
-evdi_version='1.14.7'
+evdi_version='1.14.8'
 
 EVDI_DIR=$(dirname "${BASH_SOURCE[0]}")
 EVDI_REBOOT_RATIONALE=
diff --git a/module/evdi_platform_drv.h b/module/evdi_platform_drv.h
index 1aafa8c..bd3ab31 100644
--- a/module/evdi_platform_drv.h
+++ b/module/evdi_platform_drv.h
@@ -25,11 +25,11 @@ struct platform_device_info;
 
 #define DRIVER_NAME   "evdi"
 #define DRIVER_DESC   "Extensible Virtual Display Interface"
-#define DRIVER_DATE   "20240920"
+#define DRIVER_DATE   "20241216"
 
 #define DRIVER_MAJOR 1
 #define DRIVER_MINOR 14
-#define DRIVER_PATCH 7
+#define DRIVER_PATCH 8
 
 void evdi_platform_remove_all_devices(struct device *device);
 unsigned int evdi_platform_device_count(struct device *device);

From 2da27042d5e63c1313a94c4a44b22bad1037e77c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 23 Dec 2024 10:26:14 +0000
Subject: [PATCH 25/31] Fixes GH-487: Add potentially missing vector header

---
 pyevdi/Stats.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/pyevdi/Stats.h b/pyevdi/Stats.h
index 9419a64..aaab6d9 100644
--- a/pyevdi/Stats.h
+++ b/pyevdi/Stats.h
@@ -4,6 +4,7 @@
 
 #include <functional>
 #include <chrono>
+#include <vector>
 
 #include "Buffer.h"
 #include "../library/evdi_lib.h"

From 3c6b9d2a4686e7f2e941f3c9e69e895bf2e06b96 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Thu, 2 Jan 2025 13:13:03 +0100
Subject: [PATCH 26/31] Fix key enrollment on Debian bookworm

---
 module/dkms_install.sh | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/module/dkms_install.sh b/module/dkms_install.sh
index 9af412c..8a99092 100755
--- a/module/dkms_install.sh
+++ b/module/dkms_install.sh
@@ -26,7 +26,20 @@ copy_evdi_make_log()
 enroll_secureboot_key()
 {
   if command -v mokutil >/dev/null && mokutil --sb-state | grep -i "SecureBoot enabled" > /dev/null; then
-    update-secureboot-policy --enroll-key 2> /dev/null || return
+
+    if ! update-secureboot-policy --help 2>&1 | grep "new-key" && [ -f /var/lib/dkms/mok.pub ]; then
+      # If we are in situation like in debian bookworm where dkms needs mok.pub to be imported as update-secureboot-policy does not create mok
+      # see https://github.com/dell/dkms/issues/429
+      if mokutil --test-key /var/lib/dkms/mok.pub; then
+        echo "Enter a password for Secure Boot. It will be asked again after a reboot."
+        mokutil --import /var/lib/dkms/mok.pub
+        mokutil --timeout -1 || true
+      fi
+    fi
+
+    if update-secureboot-policy --help 2>&1 | grep "enroll-key"; then
+      update-secureboot-policy --enroll-key 2> /dev/null || return
+    fi
 
     if [[ -z $EVDI_REBOOT_RATIONALE && $(mokutil --list-new | wc -l) -gt 0 ]]; then
       EVDI_REBOOT_RATIONALE="SecureBoot key was enrolled during the installation."

From 9bcfbc2ea2459aed7590eaa8a7ba2e3d72fba5d9 Mon Sep 17 00:00:00 2001
From: Jakub Prussak <jakub.prussak@synaptics.com>
Date: Fri, 20 Dec 2024 09:44:15 +0100
Subject: [PATCH 27/31] Preparation for automated note generation

---
 Jenkinsfile             | 23 ++++++++++++++++++++++-
 ci/build_against_kernel |  7 +++++++
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/Jenkinsfile b/Jenkinsfile
index b2e1c3b..1a8fe45 100644
--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -35,7 +35,11 @@ pipeline {
                DETECT_JAR_DOWNLOAD_DIR = "${env.WORKSPACE}/synopsys_download"
             }
             steps {
+            script {
                   synopsys_detect detectProperties: "--detect.project.name='Evdi' --detect.project.version.name='${env.GIT_BRANCH}' --detect.excluded.directories=tmp,bd_evdi,synopsys_download --detect.output.path='${env.WORKSPACE}/bd_evdi'", downloadStrategyOverride: [$class: 'ScriptOrJarDownloadStrategy']
+                  def buildUrl = "$BUILD_URL"
+                  env.BLACKDUCK = sh(script: "curl -Lk '${buildUrl}/consoleText' | grep 'Black Duck Project BOM:'", returnStdout: true)
+            }
             }
         }
         stage ('Build evdi-amd64.deb') {
@@ -92,10 +96,13 @@ pipeline {
             }
         }
         stage ('Build against latest rc kernel') {
-            steps {
+            steps { 
                 sh '''#!/usr/bin/env bash
                 set -e
                 ./ci/build_against_kernel --repo-ci rc'''
+                script {
+                    env.KERNELS = sh(script: "./ci/build_against_kernel --list-kernels", returnStdout: true).trim()
+                }
             }
         }
         stage ('Run KUnit tests') {
@@ -146,5 +153,19 @@ pipeline {
                   string(name: 'buildNumber', value: "${env.BUILD_NUMBER}")])
           }
         }
+        stage ( 'Create note' )
+        {
+          when { anyOf { branch 'main' } }
+          steps {
+            build (
+                job: 'PPD-POSIX/create release note/master',
+                parameters: [string(name: 'kernels', value: env.KERNELS),
+                  string(name: 'git_hash', value: env.GIT_COMMIT),
+                  string(name: 'build_number', value: env.BUILD_NUMBER),
+                  string(name: 'evdi_ver', value: env.EVDI_VERSION),
+                  string(name: 'blackduck', value: env.BLACKDUCK),
+                ])
+          }
+        }
     }
 }
diff --git a/ci/build_against_kernel b/ci/build_against_kernel
index 25cf174..1eed37b 100755
--- a/ci/build_against_kernel
+++ b/ci/build_against_kernel
@@ -88,6 +88,7 @@ get_versions_from_file() { # file
 }
 
 get_supported_linux_versions() {
+  set +e
   local repo_dir
   repo_dir="$(pwd)/$TMP/linux"
   local CC="${CC:-gcc}"
@@ -300,6 +301,12 @@ do
       exit 0
       ;;
 
+    --list-kernels)
+      MODE="REPO-CI"
+      get_supported_linux_versions
+      exit 0
+      ;;
+
     -*)
       echo "Unknown option '${1}'" >&2
       usage >&2

From 26e2fc66da169856b92607cb4cc5ff131319a324 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 27 Jan 2025 07:22:48 +0000
Subject: [PATCH 28/31] Fix evdi build on for 6.14 kernels

---
 module/evdi_drm_drv.c      | 3 +++
 module/evdi_platform_drv.h | 5 +++++
 2 files changed, 8 insertions(+)

diff --git a/module/evdi_drm_drv.c b/module/evdi_drm_drv.c
index 7959534..7a61f94 100644
--- a/module/evdi_drm_drv.c
+++ b/module/evdi_drm_drv.c
@@ -147,7 +147,10 @@ static struct drm_driver driver = {
 
 	.name = DRIVER_NAME,
 	.desc = DRIVER_DESC,
+#if KERNEL_VERSION(6, 14, 0) <= LINUX_VERSION_CODE
+#else
 	.date = DRIVER_DATE,
+#endif
 	.major = DRIVER_MAJOR,
 	.minor = DRIVER_MINOR,
 	.patchlevel = DRIVER_PATCH,
diff --git a/module/evdi_platform_drv.h b/module/evdi_platform_drv.h
index bd3ab31..b99014c 100644
--- a/module/evdi_platform_drv.h
+++ b/module/evdi_platform_drv.h
@@ -20,12 +20,17 @@
 #ifndef _EVDI_PLATFORM_DRV_H_
 #define _EVDI_PLATFORM_DRV_H_
 
+#include <linux/version.h>
+
 struct device;
 struct platform_device_info;
 
 #define DRIVER_NAME   "evdi"
 #define DRIVER_DESC   "Extensible Virtual Display Interface"
+#if KERNEL_VERSION(6, 14, 0) <= LINUX_VERSION_CODE
+#else
 #define DRIVER_DATE   "20241216"
+#endif
 
 #define DRIVER_MAJOR 1
 #define DRIVER_MINOR 14

From 7d821a30b69176ccbed5529c5f356b1d2f9fc691 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 17 Feb 2025 11:39:52 +0100
Subject: [PATCH 29/31] Check first origin/main then github/main

---
 Jenkinsfile_linux_master | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Jenkinsfile_linux_master b/Jenkinsfile_linux_master
index ff7311d..bda7d0c 100644
--- a/Jenkinsfile_linux_master
+++ b/Jenkinsfile_linux_master
@@ -16,7 +16,7 @@ pipeline {
                        git fetch github
                      '''
 
-                      def list = [ "github/main", "origin/main" ]
+                      def list = [ "origin/main", "github/main" ]
                       for(int i=0; i < list.size(); i++) {
                         def branchName = list[i];
                         stage(branchName) {

From 0f37d18f3335127a088f50c85ee1910a9e299758 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C5=81ukasz=20Spintzyk?= <Lukasz.Spintzyk@synaptics.com>
Date: Mon, 27 Jan 2025 07:33:17 +0000
Subject: [PATCH 30/31] Remove unnecessary dependencies in evdi package

libdrm-dev and libc6-dev are build dependencies of libevdi.so.
Since that library is delivered in binary form - we don't need them.
Any other dependencies of evdi are expressed via dependencies of dkms
package.
---
 ci/prepare_deb_package | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ci/prepare_deb_package b/ci/prepare_deb_package
index 5d850a3..1f15509 100755
--- a/ci/prepare_deb_package
+++ b/ci/prepare_deb_package
@@ -99,7 +99,7 @@ Version: $evdi_version-$debian_revision
 Maintainer: $evdi_maintainer
 Source: evdi
 Section: kernel
-Depends: mokutil, dkms, libdrm-dev, libc6-dev
+Depends: mokutil, dkms
 Conflicts: libevdi${libevdi_abi}, libevdi${libevdi_abi}-dev, evdi-dkms
 Architecture: $evdi_arch
 Priority: optional

From 2ca67e89a15ef01bf36abbb217fd74081faf9b4e Mon Sep 17 00:00:00 2001
From: Jakub Prussak <jakub.prussak@synaptics.com>
Date: Tue, 25 Mar 2025 08:36:12 +0100
Subject: [PATCH 31/31] Set version to 1.14.9

---
 ci/deb_config              | 2 +-
 library/Makefile           | 2 +-
 library/evdi_lib.h         | 2 +-
 module/Makefile            | 2 +-
 module/dkms.conf           | 2 +-
 module/dkms_install.sh     | 2 +-
 module/evdi_platform_drv.h | 4 ++--
 7 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/ci/deb_config b/ci/deb_config
index 86ef144..4cb0b66 100644
--- a/ci/deb_config
+++ b/ci/deb_config
@@ -8,7 +8,7 @@ evdi_package_name='evdi'
 evdi_maintainer='Synaptics Technical Support <technical-enquiries@synaptics.com>'
 
 # shellcheck disable=SC2034
-evdi_version='1.14.8'
+evdi_version='1.14.9'
 
 # shellcheck disable=SC2034
 evdi_description="Extensible Virtual Display Interface (EVDI) is a Linux® kernel module that enables management of multiple screens.
diff --git a/library/Makefile b/library/Makefile
index 0401878..8fd4f46 100644
--- a/library/Makefile
+++ b/library/Makefile
@@ -11,7 +11,7 @@ PKG_CONFIG ?= pkg-config
 DEPS = evdi_ioctl.h
 CFLAGS := -I../module -std=gnu99 -fPIC -D_FILE_OFFSET_BITS=64 $(CFLAGS) $$($(PKG_CONFIG) --cflags-only-I libdrm)
 
-LIBVER := 1.14.8
+LIBVER := 1.14.9
 LIBABI := 1
 
 PREFIX ?= /usr/local
diff --git a/library/evdi_lib.h b/library/evdi_lib.h
index 05c3a65..a19c564 100644
--- a/library/evdi_lib.h
+++ b/library/evdi_lib.h
@@ -15,7 +15,7 @@ extern "C" {
 
 #define LIBEVDI_VERSION_MAJOR 1
 #define LIBEVDI_VERSION_MINOR 14
-#define LIBEVDI_VERSION_PATCH 8
+#define LIBEVDI_VERSION_PATCH 9
 
 struct evdi_lib_version {
 	int version_major;
diff --git a/module/Makefile b/module/Makefile
index 4bbe446..b354d8f 100644
--- a/module/Makefile
+++ b/module/Makefile
@@ -63,7 +63,7 @@ CP ?= cp
 DKMS ?= dkms
 RM ?= rm
 
-MODVER=1.14.8
+MODVER=1.14.9
 
 ifeq ($(KVER),)
 	KVER := $(shell uname -r)
diff --git a/module/dkms.conf b/module/dkms.conf
index 93983bc..5e4da64 100644
--- a/module/dkms.conf
+++ b/module/dkms.conf
@@ -7,7 +7,7 @@
 #
 
 PACKAGE_NAME="evdi"
-PACKAGE_VERSION=1.14.8
+PACKAGE_VERSION=1.14.9
 AUTOINSTALL=yes
 
 MAKE[0]="make all INCLUDEDIR=/lib/modules/$kernelver/build/include KVERSION=$kernelver DKMS_BUILD=1"
diff --git a/module/dkms_install.sh b/module/dkms_install.sh
index 8a99092..c9f2892 100755
--- a/module/dkms_install.sh
+++ b/module/dkms_install.sh
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: GPL-2.0-only
 # Copyright (c) 2023 DisplayLink (UK) Ltd.
 
-evdi_version='1.14.8'
+evdi_version='1.14.9'
 
 EVDI_DIR=$(dirname "${BASH_SOURCE[0]}")
 EVDI_REBOOT_RATIONALE=
diff --git a/module/evdi_platform_drv.h b/module/evdi_platform_drv.h
index b99014c..b71ad41 100644
--- a/module/evdi_platform_drv.h
+++ b/module/evdi_platform_drv.h
@@ -29,12 +29,12 @@ struct platform_device_info;
 #define DRIVER_DESC   "Extensible Virtual Display Interface"
 #if KERNEL_VERSION(6, 14, 0) <= LINUX_VERSION_CODE
 #else
-#define DRIVER_DATE   "20241216"
+#define DRIVER_DATE   "20250325"
 #endif
 
 #define DRIVER_MAJOR 1
 #define DRIVER_MINOR 14
-#define DRIVER_PATCH 8
+#define DRIVER_PATCH 9
 
 void evdi_platform_remove_all_devices(struct device *device);
 unsigned int evdi_platform_device_count(struct device *device);
