diff -ru evdi-1.12.0-a/module/evdi_drm_drv.h evdi-1.12.0-b/module/evdi_drm_drv.h
--- evdi-1.12.0-a/module/evdi_drm_drv.h	2022-11-25 12:52:30.000000000 -0500
+++ evdi-1.12.0-b/module/evdi_drm_drv.h	2022-11-25 13:41:41.628392700 -0500
@@ -25,7 +25,7 @@
 #else
 #include <drm/drmP.h>
 #endif
-#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL9)
+#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 #include <drm/drm_framebuffer.h>
 #else
 #include <drm/drm_irq.h>
diff -ru evdi-1.12.0-a/module/evdi_gem.c evdi-1.12.0-b/module/evdi_gem.c
--- evdi-1.12.0-a/module/evdi_gem.c	2022-11-25 12:52:30.000000000 -0500
+++ evdi-1.12.0-b/module/evdi_gem.c	2022-11-25 13:51:11.464137300 -0500
@@ -10,7 +10,7 @@
 
 #include <linux/sched.h>
 #include <linux/version.h>
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 #elif KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE
 #include <linux/dma-buf-map.h>
 #endif
@@ -291,7 +291,7 @@
 	int ret;
 
 	if (evdi_drm_gem_object_use_import_attach(&obj->base)) {
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 		struct iosys_map map = IOSYS_MAP_INIT_VADDR(NULL);
 #elif KERNEL_VERSION(5, 11, 0) <= LINUX_VERSION_CODE || defined(EL8)
 		struct dma_buf_map map = DMA_BUF_MAP_INIT_VADDR(NULL);
@@ -324,7 +324,7 @@
 void evdi_gem_vunmap(struct evdi_gem_object *obj)
 {
 	if (evdi_drm_gem_object_use_import_attach(&obj->base)) {
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 		struct iosys_map map = IOSYS_MAP_INIT_VADDR(NULL);
 
 		if (obj->vmap_is_iomem)
Only in evdi-1.12.0-b/module: evdi_gem.c.oqBqCmy
diff -ru evdi-1.12.0-a/module/evdi_ioc32.c evdi-1.12.0-b/module/evdi_ioc32.c
--- evdi-1.12.0-a/module/evdi_ioc32.c	2022-11-25 12:52:30.000000000 -0500
+++ evdi-1.12.0-b/module/evdi_ioc32.c	2022-11-25 13:41:41.662675200 -0500
@@ -62,9 +62,9 @@
 	if (copy_from_user(&req32, (void __user *)arg, sizeof(req32)))
 		return -EFAULT;
 
-#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL9)
+#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 	request = kmalloc(sizeof(*request), GFP_USER);
-#elif KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE || defined(EL8)
+#elif KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE
 	request = compat_alloc_user_space(sizeof(*request));
 #endif
 #if KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE || defined(EL8)
@@ -95,9 +95,9 @@
 	if (copy_from_user(&req32, (void __user *)arg, sizeof(req32)))
 		return -EFAULT;
 
-#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL9)
+#if KERNEL_VERSION(5, 15, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
 	request = kmalloc(sizeof(*request), GFP_USER);
-#elif KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE || defined(EL8)
+#elif KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE
 	request = compat_alloc_user_space(sizeof(*request));
 #endif
 #if KERNEL_VERSION(5, 0, 0) <= LINUX_VERSION_CODE || defined(EL8)
