diff -uNr evdi-1.14.7-a/module/evdi_gem.c evdi-1.14.7-b/module/evdi_gem.c
--- evdi-1.14.7-a/module/evdi_gem.c	2024-12-12 01:17:29.000000000 +0000
+++ evdi-1.14.7-b/module/evdi_gem.c	2024-12-12 01:21:31.712582043 +0000
@@ -29,6 +29,8 @@
 #include <linux/vmalloc.h>
 
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
+#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
 
diff -uNr evdi-1.14.7-a/module/evdi_painter.c evdi-1.14.7-b/module/evdi_painter.c
--- evdi-1.14.7-a/module/evdi_painter.c	2024-12-12 01:17:29.000000000 +0000
+++ evdi-1.14.7-b/module/evdi_painter.c	2024-12-12 01:22:38.447833469 +0000
@@ -37,6 +37,8 @@
 
 /* Import of DMA_BUF namespace was reverted in EL8 */
 #if KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
+MODULE_IMPORT_NS(DMA_BUF);
+#elif KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #endif
 
