diff -urN evdi-1.14.11-a/module/evdi_gem.c evdi-1.14.11-b/module/evdi_gem.c
--- evdi-1.14.11-a/module/evdi_gem.c	2025-08-26 08:31:36.000000000 +0000
+++ evdi-1.14.11-b/module/evdi_gem.c	2025-09-29 13:02:14.934235284 +0000
@@ -29,7 +29,7 @@
 #include <linux/vmalloc.h>
 
 
-#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE || defined(EL10)
+#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #elif KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
 MODULE_IMPORT_NS(DMA_BUF);
diff -urN evdi-1.14.11-a/module/evdi_painter.c evdi-1.14.11-b/module/evdi_painter.c
--- evdi-1.14.11-a/module/evdi_painter.c	2025-08-26 08:31:36.000000000 +0000
+++ evdi-1.14.11-b/module/evdi_painter.c	2025-09-29 13:01:30.620399278 +0000
@@ -36,7 +36,7 @@
 #endif
 
 /* Import of DMA_BUF namespace was reverted in EL8 */
-#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE || defined(EL10)
+#if KERNEL_VERSION(6, 13, 0) <= LINUX_VERSION_CODE
 MODULE_IMPORT_NS("DMA_BUF");
 #elif KERNEL_VERSION(5, 16, 0) <= LINUX_VERSION_CODE || defined(EL9)
 MODULE_IMPORT_NS(DMA_BUF);
