diff -urN tmp-a/ci/deb_config tmp-b/ci/deb_config
--- tmp-a/ci/deb_config	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/ci/deb_config	2026-04-28 05:33:57.000000000 -0400
@@ -8,7 +8,7 @@
 evdi_maintainer='Synaptics Technical Support <technical-enquiries@synaptics.com>'
 
 # shellcheck disable=SC2034
-evdi_version='1.14.15'
+evdi_version='1.14.16'
 
 # shellcheck disable=SC2034
 evdi_description="Extensible Virtual Display Interface (EVDI) is a Linux® kernel module that enables management of multiple screens.
diff -urN tmp-a/library/evdi_lib.h tmp-b/library/evdi_lib.h
--- tmp-a/library/evdi_lib.h	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/library/evdi_lib.h	2026-04-28 05:33:57.000000000 -0400
@@ -15,7 +15,7 @@
 
 #define LIBEVDI_VERSION_MAJOR 1
 #define LIBEVDI_VERSION_MINOR 14
-#define LIBEVDI_VERSION_PATCH 15
+#define LIBEVDI_VERSION_PATCH 16
 
 struct evdi_lib_version {
 	int version_major;
diff -urN tmp-a/library/Makefile tmp-b/library/Makefile
--- tmp-a/library/Makefile	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/library/Makefile	2026-04-28 05:33:57.000000000 -0400
@@ -11,7 +11,7 @@
 DEPS = evdi_ioctl.h
 CFLAGS := -I../module -std=gnu99 -fPIC -D_FILE_OFFSET_BITS=64 $(CFLAGS) $$($(PKG_CONFIG) --cflags-only-I libdrm)
 
-LIBVER := 1.14.15
+LIBVER := 1.14.16
 LIBABI := 1
 
 PREFIX ?= /usr
diff -urN tmp-a/module/dkms.conf tmp-b/module/dkms.conf
--- tmp-a/module/dkms.conf	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/dkms.conf	2026-04-28 05:33:57.000000000 -0400
@@ -7,7 +7,7 @@
 #
 
 PACKAGE_NAME="evdi"
-PACKAGE_VERSION=1.14.15
+PACKAGE_VERSION=1.14.16
 AUTOINSTALL=yes
 
 MAKE[0]="make all INCLUDEDIR=/lib/modules/$kernelver/build/include KVERSION=$kernelver DKMS_BUILD=1"
diff -urN tmp-a/module/dkms_install.sh tmp-b/module/dkms_install.sh
--- tmp-a/module/dkms_install.sh	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/dkms_install.sh	2026-04-28 05:33:57.000000000 -0400
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: GPL-2.0-only
 # Copyright (c) 2023 DisplayLink (UK) Ltd.
 
-evdi_version='1.14.15'
+evdi_version='1.14.16'
 
 EVDI_DIR=$(dirname "${BASH_SOURCE[0]}")
 EVDI_REBOOT_RATIONALE=
diff -urN tmp-a/module/evdi_drm_drv.h tmp-b/module/evdi_drm_drv.h
--- tmp-a/module/evdi_drm_drv.h	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/evdi_drm_drv.h	2026-04-28 05:33:57.000000000 -0400
@@ -102,7 +102,7 @@
 struct drm_framebuffer *evdi_fb_user_fb_create(
 				struct drm_device *dev,
 				struct drm_file *file,
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 				const struct drm_format_info *info,
 #endif
 				const struct drm_mode_fb_cmd2 *mode_cmd);
diff -urN tmp-a/module/evdi_fb.c tmp-b/module/evdi_fb.c
--- tmp-a/module/evdi_fb.c	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/evdi_fb.c	2026-04-28 05:33:57.000000000 -0400
@@ -203,20 +203,20 @@
 static int
 evdi_framebuffer_init(struct drm_device *dev,
 		      struct evdi_framebuffer *efb,
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 		      const struct drm_format_info *info,
 #endif
 		      const struct drm_mode_fb_cmd2 *mode_cmd,
 		      struct evdi_gem_object *obj)
 {
 	efb->obj = obj;
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 	if (info == NULL)
 		info = drm_get_format_info(dev, mode_cmd->pixel_format,
 					   mode_cmd->modifier[0]);
 #endif
 	drm_helper_mode_fill_fb_struct(dev, &efb->base,
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 				       info,
 #endif
 				       mode_cmd);
@@ -253,7 +253,7 @@
 struct drm_framebuffer *evdi_fb_user_fb_create(
 					struct drm_device *dev,
 					struct drm_file *file,
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 					const struct drm_format_info *info,
 #endif
 					const struct drm_mode_fb_cmd2 *mode_cmd)
@@ -289,7 +289,7 @@
 	efb->base.obj[0] = obj;
 
 	ret = evdi_framebuffer_init(dev, efb,
-#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE
+#if KERNEL_VERSION(6, 17, 0) <= LINUX_VERSION_CODE || defined(CENTOS9) || defined(CENTOS10)
 				    info,
 #endif
 				    mode_cmd, to_evdi_bo(obj));
diff -urN tmp-a/module/evdi_gem.c tmp-b/module/evdi_gem.c
--- tmp-a/module/evdi_gem.c	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/evdi_gem.c	2026-04-28 05:33:57.000000000 -0400
@@ -539,7 +539,7 @@
 
 	npages = DIV_ROUND_UP(attach->dmabuf->size, PAGE_SIZE);
 	npages_sg = evdi_get_sg_table_pages_num(sg);
-	npages = npages < npages_sg ? npages_sg : npages_sg;
+	npages = max(npages, npages_sg);
 	DRM_DEBUG_PRIME("Importing %d pages\n", npages);
 	obj->pages = kvmalloc_array(npages, sizeof(struct page *), GFP_KERNEL);
 	if (!obj->pages) {
diff -urN tmp-a/module/evdi_painter.c tmp-b/module/evdi_painter.c
--- tmp-a/module/evdi_painter.c	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/evdi_painter.c	2026-04-28 05:33:57.000000000 -0400
@@ -173,7 +173,7 @@
 	*count = 1;
 }
 
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if (KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)) && defined(CONFIG_X86)
 static int copy_primary_pixels_on_xe(struct evdi_framebuffer *efb,
 			       char __user *buffer,
 			       int buf_byte_stride,
@@ -185,6 +185,9 @@
 	const int byte_span = max_x * 4;
 	struct iosys_map dst_mapping = IOSYS_MAP_INIT_VADDR(vmalloc(max_x * 4));
 
+	if (!dst_mapping.vaddr)
+		return -ENOMEM;
+
 	for (y = 0; y < max_y; ++y) {
 		const int src_offset = fb->offsets[0] + fb->pitches[0] * y;
 		struct iosys_map src_mapping = IOSYS_MAP_INIT_VADDR((char *)efb->obj->vmapping + src_offset);
@@ -193,8 +196,10 @@
 
 		drm_clflush_virt_range(src_mapping.vaddr, byte_span);
 		drm_memcpy_from_wc(&dst_mapping, &src_mapping, byte_span);
-		if (copy_to_user(dst, dst_mapping.vaddr, byte_span))
+		if (copy_to_user(dst, dst_mapping.vaddr, byte_span)) {
+			vfree(dst_mapping.vaddr);
 			return -EFAULT;
+		}
 
 		src_mapping.vaddr += fb->pitches[0];
 		dst += buf_byte_stride;
@@ -218,7 +223,7 @@
 
 	EVDI_CHECKPT();
 
-#if KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)
+#if (KERNEL_VERSION(5, 18, 0) <= LINUX_VERSION_CODE || defined(EL8) || defined(EL9)) && defined(CONFIG_X86)
 	if (efb->is_from_xe)
 		return copy_primary_pixels_on_xe(efb, buffer, buf_byte_stride, max_x, max_y);
 #endif
diff -urN tmp-a/module/evdi_platform_drv.h tmp-b/module/evdi_platform_drv.h
--- tmp-a/module/evdi_platform_drv.h	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/evdi_platform_drv.h	2026-04-28 05:33:57.000000000 -0400
@@ -29,12 +29,12 @@
 #define DRIVER_DESC   "Extensible Virtual Display Interface"
 #if KERNEL_VERSION(6, 14, 0) <= LINUX_VERSION_CODE || defined(EL9) || defined(EL10)
 #else
-#define DRIVER_DATE   "20260225"
+#define DRIVER_DATE   "20260428"
 #endif
 
 #define DRIVER_MAJOR 1
 #define DRIVER_MINOR 14
-#define DRIVER_PATCH 15
+#define DRIVER_PATCH 16
 
 void evdi_platform_remove_all_devices(struct device *device);
 unsigned int evdi_platform_device_count(struct device *device);
diff -urN tmp-a/module/Makefile tmp-b/module/Makefile
--- tmp-a/module/Makefile	2026-02-25 02:24:01.000000000 -0500
+++ tmp-b/module/Makefile	2026-04-28 05:33:57.000000000 -0400
@@ -12,6 +12,10 @@
 ELFLAG := -DEL$(shell echo $(VERSION_ID) | cut -d. -f1)
 endif
 
+ifneq (,$(findstring centos,$(ID)))
+CENTOSFLAG := -DCENTOS$(shell echo $(VERSION_ID) | cut -d. -f1)
+endif
+
 Raspbian := $(shell grep -Eic 'raspb(erry|ian)' /proc/cpuinfo /etc/os-release 2>/dev/null )
 ifeq (,$(findstring 0, $(Raspbian)))
 RPIFLAG := -DRPI
@@ -23,7 +27,7 @@
 
 KERN_DIR := /lib/modules/$(KERNELRELEASE)/build
 
-ccflags-y := -Iinclude/uapi/drm -Iinclude/drm $(ELFLAG) $(RPIFLAG)
+ccflags-y := -Iinclude/uapi/drm -Iinclude/drm $(ELFLAG) $(CENTOSFLAG) $(RPIFLAG)
 evdi-y := evdi_platform_drv.o evdi_platform_dev.o evdi_sysfs.o evdi_modeset.o evdi_connector.o evdi_encoder.o evdi_drm_drv.o evdi_fb.o evdi_gem.o evdi_painter.o evdi_params.o evdi_cursor.o evdi_debug.o evdi_i2c.o
 evdi-$(CONFIG_COMPAT) += evdi_ioc32.o
 obj-m := evdi.o
@@ -45,7 +49,7 @@
 
 # inside kbuild
 # Note: this can be removed once it is in kernel tree and Kconfig is properly used
-ccflags-y := -isystem include/uapi/drm $(CFLAGS) $(ELFLAG) $(RPIFLAG)
+ccflags-y := -isystem include/uapi/drm $(CFLAGS) $(ELFLAG) $(CENTOSFLAG) $(RPIFLAG)
 evdi-y := evdi_platform_drv.o evdi_platform_dev.o evdi_sysfs.o evdi_modeset.o evdi_connector.o evdi_encoder.o evdi_drm_drv.o evdi_fb.o evdi_gem.o evdi_painter.o evdi_params.o evdi_cursor.o evdi_debug.o evdi_i2c.o
 evdi-$(CONFIG_COMPAT) += evdi_ioc32.o
 CONFIG_DRM_EVDI ?= m
@@ -59,7 +63,7 @@
 DKMS ?= dkms
 RM ?= rm
 
-MODVER=1.14.15
+MODVER=1.14.16
 
 ifeq ($(KVER),)
 	KVER := $(shell uname -r)
